### ISSUE TITLE
Learn CockroachDB

### ISSUE THREAD

**Issue by leandrosardi** [leandrosardi/free-membership-sites#3]
## Goals

1. instalar CockrouchDB on AWS ES2
2. conectar la base de datos en AWS ES2
3. acceder el dashboard cockrouchDB
4. instalar esquema de base de datos en cockrouch DB 

## Research

Bassicaly, we need to know:

1. Is it absolutely free like PostgreSQL?
2. Can I run it on my own servers, or I have to host my databases on its site?
3. Is it a layer running over PostgreSQL that just take care of sharding and mirroring? Or it has its own RDBMS engine?

Aside that CockrouchDB allows us an elastic database scalability, we need a procedure for DBA.

**Comment by leandrosardi** [comment]
## Documentation I found:

### How to connect postgresSQL from Sequel:

- https://stackoverflow.com/questions/12134596/how-do-i-connect-to-a-postgres-database-with-sequel

- https://sequel.jeremyevans.net/rdoc/files/doc/opening_databases_rdoc.html

Also, I would like to connect using TinyTDS

### How to do Database Sharding in PostgreSQL

- https://www.google.com/search?q=posgres+table+sharding&oq=posgres+table+sharding&aqs=chrome..69i57j0i13j0i390.11747j0j4&client=ms-android-om-lge&sourceid=chrome-mobile&ie=UTF-8

https://about.gitlab.com/handbook/engineering/development/enablement/database/doc/fdw-sharding.html

### Research about CockRoachDB

- https://github.com/cockroachdb/cockroach

- https://www.cockroachlabs.com/docs/stable/start-a-local-cluster.html

- https://www.cockroachlabs.com/docs/stable/install-cockroachdb.html

**Comment by leandrosardi** [comment]
> - https://www.cockroachlabs.com/docs/stable/install-cockroachdb.html
> - https://www.cockroachlabs.com/docs/v21.2/start-a-local-cluster

I installed the database, but I can´t access the dashboard from the internet.

**Comment by leandrosardi** [comment]
I am facing 2 problems running CockroachDB on AWS/EC2:

## Problem 1: I can't access the dashboard from the Internet.

1. I followed instructions in these 2 tutorials

- https://www.cockroachlabs.com/docs/stable/install-cockroachdb.html

- https://www.cockroachlabs.com/docs/v21.2/start-a-local-cluster

2. I also watched this videoÑ

- https://www.youtube.com/watch?v=wUaGyUpyE2E

3. I added port 8080 to the firewall of AWS-EC2

![image](https://user-images.githubusercontent.com/55877846/169600671-a65e9281-f247-4fef-964e-c616892dd696.png)

4. I ran this command:

![image](https://user-images.githubusercontent.com/55877846/169600392-57bb076f-f778-4a82-a40f-d33a7d483502.png)

5. But I can`t access the dashboard from my computer:

![image](https://user-images.githubusercontent.com/55877846/169600884-83ee9867-02a8-489d-988a-36eb61cd8074.png)

**Comment by leandrosardi** [comment]
## Problem 2: I can´t connect the database from the Internet

1. I grabed the connection parameters from the installation of CockrouchDB:

![image](https://user-images.githubusercontent.com/55877846/169601180-91d5243a-2bd5-4bd2-8576-08c0b3f5fbba.png)

2. I am adding new server with these parameters on pgAdmin in my computer:

![image](https://user-images.githubusercontent.com/55877846/169601293-9eaea730-3188-479d-a9ca-b7359727514b.png)

**Comment by leandrosardi** [comment]
## Email sent to CockrouchDB support

> We need to install CockrouchDB on a AWS-EC2 instance.
> 
> We are facing problems for:
> 1. access the web dashboard of CockrouchDB from the web.
> 2. access the postgreSQL database from the web.
> 
> More details here:
> 
> - https://github.com/leandrosardi/free-membership-sites/issues/3#issuecomment-1133271888
> 
> - https://github.com/leandrosardi/free-membership-sites/issues/3#issuecomment-1133275751

## Response received from CockrouchDB support

> Hi,
> 
> From the screenshots you provided, it looks like you are running cockroach demo. That only allows network connections on localhost (127.0.0.1), it doesn't allow external access as it's just a demonstration tool.
> 
> You need to run through the setup details for cockroach start (https://www.cockroachlabs.com/docs/stable/cockroach-start.html).
> 
> This page demonstrates how to set up a secure cluster: https://www.cockroachlabs.com/docs/stable/secure-a-cluster.html
> 
> Alternatively, you can use CockroachCloud (https://cockroachlabs.cloud/) to set up a free Serverless cluster or a paid-for Dedicated cluster.
> 
> Kind Regards,
> Shaun

**Comment by leandrosardi** [comment]
After reading documentation, I built this procedure to install a production multi-node cluster:

## Step 1. Install CockroachDB
Reference: https://www.cockroachlabs.com/docs/stable/install-cockroachdb.html

1. Download the binary
```
curl https://binaries.cockroachdb.com/cockroach-v21.2.10.linux-amd64.tgz | tar -xz && sudo cp -i cockroach-v21.2.10.linux-amd64/cockroach /usr/local/bin/
```

2. Create the directory where the external libraries will be stored:
```
mkdir -p /usr/local/lib/cockroach
```

3. Copy the library files to the directory:
```
cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos.so /usr/local/lib/cockroach/
cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos_c.so /usr/local/lib/cockroach/
```

4. Verify that CockroachDB can execute spatial queries.
```
which cockroach
```

## Step 2. Secure a Cluster
Reference: https://www.cockroachlabs.com/docs/stable/secure-a-cluster.html

1. Create two directories:
```
mkdir certs my-safe-directory
```

2. Create the CA (Certificate Authority) certificate and key pair:
```
cockroach cert create-ca --certs-dir=certs --ca-key=my-safe-directory/ca.key
```

3. Create the certificate and key pair for your nodes:
```
cockroach cert create-node localhost $(hostname) --certs-dir=certs --ca-key=my-safe-directory/ca.key
```

Because you're running a local cluster and all nodes use the same hostname (localhost), you only need a single node certificate. 
Note that this is different than running a production cluster, where you would need to generate a certificate and key for each node, issued to all common names and IP addresses you might use to refer to the node as well as to any load balancer instances.


4. Create a client certificate and key pair for the root user:
```
cockroach cert create-client root --certs-dir=certs --ca-key=my-safe-directory/ca.key
```


## Start 3. Start a Production Cluseter
Reference: https://www.cockroachlabs.com/docs/stable/cockroach-start.html

1. start a new multi-node cluster or add nodes to an existing cluster.

```
cockroach start --certs-dir=certs --store=node0 --listen-addr=localhost:26200 --http-addr=54.86.27.191:8000 --join=localhost:26200,localhost:26201,localhost:26202 --background
```

**Comment by leandrosardi** [comment]
> start a new multi-node cluster or add nodes to an existing cluster.
> cockroach start --certs-dir=certs --store=node0 --listen-addr=localhost:26200 --http-addr=54.86.27.191:8000 --join=localhost:26200,localhost:26201,localhost:26202 --background

I am getting this error:

```
*
* ERROR: ERROR: cannot dial server.
* Is the server running?
* If the server is running, check --host client-side and --advertise server-side.
*
* cockroach server exited with error: consider changing the port via --http-addr: listen tcp 54.86.27.191:8000: bind: cannot assign requested address
*
ERROR: cannot dial server.
Is the server running?
If the server is running, check --host client-side and --advertise server-side.

cockroach server exited with error: consider changing the port via --http-addr: listen tcp 54.86.27.191:8000: bind: cannot assign requested address
Failed running "start"
E220523 13:57:55.828297 1 1@cli/clierror/check.go:35  [-] 1  ERROR: exit status 1
ERROR: exit status 1
Failed running "start"
```

**Comment by leandrosardi** [comment]
> https://www.cockroachlabs.com/docs/stable/cockroach-start.html

Response from support:

> This suggests that the IP address (54.86.27.191) isn't available as an interface on this machine. Are you perhaps trying to assign the public IP address to it? It should be an address available from the ip addr command.

So, I ran this:

```
ubuntu@ip-172-31-22-97:~$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc fq_codel state UP group default qlen 1000
    link/ether 0a:13:e4:53:83:3d brd ff:ff:ff:ff:ff:ff
    inet 172.31.22.97/20 metric 100 brd 172.31.31.255 scope global dynamic eth0
       valid_lft 1932sec preferred_lft 1932sec
    inet6 fe80::813:e4ff:fe53:833d/64 scope link
       valid_lft forever preferred_lft forever
ubuntu@ip-172-31-22-97:~$ cockroach start --certs-dir=certs --store=node10 --listen-addr=localhost:26300 --http-addr=172.31.22.97:8001 --join=localhost:26300,localhost:26301,localhost:26302 --background
*
* INFO: initial startup completed.
* Node will now attempt to join a running cluster, or wait for `cockroach init`.
* Client connections will be accepted after this completes successfully.
* Check the log file(s) for progress.
*
ubuntu@ip-172-31-22-97:~$
```

**Comment by leandrosardi** [comment]
> But now I am getting this 404 - Page Not Found problem:

I received this response form support.

> Hi,
> 
> It doesn't look like you initialised the cluster. This is an operation you have to do for the very first time. As a result, you have 'uninitialised nodes' seeking to join an existing cluster, but no existing cluster exists, so you have to explicitly tell the nodes to create a fresh cluster. The command is cockroach init.

I ran such a command, but I am getting this error, when I installed the certificates as is detailed in the tutorial:

![image](https://user-images.githubusercontent.com/55877846/169908535-9353b1df-5619-4df5-bdcb-22e39a934a80.png)

**Comment by leandrosardi** [comment]
> I ran such a command, but I am getting this error, when I installed the certificates as is detailed in the tutorial:

Reply from support:

> Hi,
> 
> You must still provide any certificate and connection strings that you normally would add. I can see from the screenshot that you have your certificate directory in a non-default location (--certs-dir=certs). So you need to also add that to the cockroach init command.
> 
> The currently running cockroach init command can be safely cancelled.
> 
> Kind Regards,
> Shaun

I added such a parameters (--certs-dir=certs).

Now I am getting this other error message:

![image](https://user-images.githubusercontent.com/55877846/169909343-8feb6a4c-f0e4-41fc-b994-f9dfbfcea73a.png)

**Comment by leandrosardi** [comment]
Reply from support:

> It looks like you need to also pass --host=localhost:26300. That should initialise it. Full command would be cockroach init --host=localhost:26300 --certs-dir=certs.
> 
> Are you running all the nodes on the same server? It appears that you are. You will run into difficulties if you use localhost as a listening address across multiple nodes as they won't be contactable by any other node. However, for a test on the same machine, it should be fine.

It is working now, as you can see in the picture below.

Regarding this:
> Are you running all the nodes on the same server? It appears that you are. You will run into difficulties if you use localhost as a listening address across multiple nodes as they won't be contactable by any other node. However, for a test on the same machine, it should be fine.

I am just running the first node. I plan to run other 2 nodes on 2 different servers.
So, I have to run this:

```
cockroach start --certs-dir=certs --store=node10 --listen-addr=172.31.22.97:26300 --http-addr=172.31.22.97:8001 --join=172.31.22.97:26300,<ip address of the second server>:26301,<ip address of the thirth server>:26302 --background
```

instead of this

```
cockroach start --certs-dir=certs --store=node10 --listen-addr=localhost:26300 --http-addr=172.31.22.97:8001 --join=localhost:26300,localhost:26301,localhost:26302 --background
```

right?


![image](https://user-images.githubusercontent.com/55877846/169911199-0d936196-8aba-451f-a2d6-076ebd70195f.png)

**Comment by leandrosardi** [comment]
> So, I have to run this:
> instead of this
> right?

I received this reply from support:

> Hi,
> 
> Yes, you will have to use network-reachable address and not use localhost as other servers would not be able to contact those nodes. You can also use the same port numbers if you're using multiple servers so you can simply your commands by not having to specify the port number.
> 
> When using multiple servers, you could also leave out the http-addr and let it take the default value (which is the listen address but on port 8080). Also, because you specified non-default ports, you will always have to pass a --host value in commands such as cockroach init and cockroach sql. You can simplify this by not specifying ports in your start command so it reverts to the defaults, then you won't have to specify anything in any other command to talk to the cluster.
> 
> Kind Regards,
> Shaun

**Comment by leandrosardi** [comment]
I am facing this issue:

![image](https://user-images.githubusercontent.com/55877846/170068028-eff6f3be-2df4-495b-9c21-81dc5d91aaed.png)

**Comment by leandrosardi** [comment]
> Does both the port and address match the one you specified in the --listen-addr option? Is this port open through any firewalls?

Yes and Yes.

```
ubuntu@ip-172-31-22-97:~$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc fq_codel state UP group default qlen 1000
    link/ether 0a:13:e4:53:83:3d brd ff:ff:ff:ff:ff:ff
    inet 172.31.22.97/20 metric 100 brd 172.31.31.255 scope global dynamic eth0
       valid_lft 1932sec preferred_lft 1932sec
    inet6 fe80::813:e4ff:fe53:833d/64 scope link
       valid_lft forever preferred_lft forever
ubuntu@ip-172-31-22-97:~$ cockroach start --certs-dir=certs --store=node10 --listen-addr=localhost:26300 --http-addr=172.31.22.97:8001 --join=localhost:26300,localhost:26301,localhost:26302 --background
*
* INFO: initial startup completed.
* Node will now attempt to join a running cluster, or wait for `cockroach init`.
* Client connections will be accepted after this completes successfully.
* Check the log file(s) for progress.
*
ubuntu@ip-172-31-22-97:~$
```

![image](https://user-images.githubusercontent.com/55877846/170073033-3813d001-a6b0-47a7-8b3b-a7157b3cf85d.png)

![image](https://user-images.githubusercontent.com/55877846/170073064-2cdf6060-6e57-44bc-acf7-887eba4e9977.png)

**Comment by leandrosardi** [comment]
> It looks like you started the node on localhost, and when you specified the host in cockroach init, you also specified localhost, so that works.

However, in the cockroach sql command, you try and specify an IP address that CockroachDB isn't listening on. If you are going to specify localhost in the listen-addr flag, you must also specify localhost in the cockroach sql command.

**Comment by leandrosardi** [comment]
> While CockroachDB has many compabilities with Postgres.

I wonder if SQL syntax is exactly the same?

What are exactly the differences between postgresql and cockroachdb ?

**Comment by leandrosardi** [comment]
> what is a good IDE to replace pgAdmin ?

**Comment by leandrosardi** [comment]
> Thank you shaun.
> 
> I am all set to start by now.
> 
> Other questions I have are:
> 
> 1. how to add a node dynamically?
> 
> 2. how to remove a node dynamically?
> 
> But I will get back on this later this week.

**Comment by leandrosardi** [comment]
Here we go again.

# CockroachDB Production Installation Procedure

After reading documentation, I built this procedure to install a production multi-node cluster:

## Step 1. Install CockroachDB
Reference: https://www.cockroachlabs.com/docs/stable/install-cockroachdb.html

1. Download the binary
```
curl https://binaries.cockroachdb.com/cockroach-v21.2.10.linux-amd64.tgz | tar -xz && sudo cp -i cockroach-v21.2.10.linux-amd64/cockroach /usr/local/bin/
```

2. Create the directory where the external libraries will be stored:
```
mkdir -p /usr/local/lib/cockroach
```

3. Copy the library files to the directory:
```
cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos.so /usr/local/lib/cockroach/
cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos_c.so /usr/local/lib/cockroach/
```

4. Verify that CockroachDB can execute spatial queries.
```
which cockroach
```

## Step 2. Secure a Cluster
Reference: https://www.cockroachlabs.com/docs/stable/secure-a-cluster.html

1. Create two directories:
```
mkdir certs my-safe-directory
```

2. Create the CA (Certificate Authority) certificate and key pair:
```
cockroach cert create-ca --certs-dir=certs --ca-key=my-safe-directory/ca.key
```

3. Create the certificate and key pair for your nodes:
```
cockroach cert create-node localhost $(hostname) --certs-dir=certs --ca-key=my-safe-directory/ca.key
```

Because you're running a local cluster and all nodes use the same hostname (localhost), you only need a single node certificate. 
Note that this is different than running a production cluster, where you would need to generate a certificate and key for each node, issued to all common names and IP addresses you might use to refer to the node as well as to any load balancer instances.


4. Create a client certificate and key pair for the root user:
```
cockroach cert create-client root --certs-dir=certs --ca-key=my-safe-directory/ca.key
```


## Start 3. Start a Production Cluseter
Reference: https://www.cockroachlabs.com/docs/stable/cockroach-start.html

1. start a new multi-node cluster or add nodes to an existing cluster.

```
cockroach start --certs-dir=certs --store=node0 --listen-addr=localhost:26200 --http-addr=54.86.27.191:8000 --join=localhost:26200,localhost:26201,localhost:26202 --background
```

## Step 4. Init the node

## Step 5. Connect the database in the same node

## Step 6. Create a user to access the dashboard and database remotely.

**Comment by leandrosardi** [comment]
Question:

I read somewhere that is not a good practice to start nodes with the --background parameter in production environments:

```
cockroach start --certs-dir=certs --store=node0 --listen-addr=<ipaddr>:26200 --http-addr=<ipaddr>:8080 --join=<ipaddr>:26200 --background
```

So, who to run the command above then?

**Comment by leandrosardi** [comment]
# Question

I have an AWS / EC2 server.

Its public IP is 54.160.160.154.
![image](https://user-images.githubusercontent.com/55877846/170106808-ba50c44d-04ad-4808-bc9b-5e0e859f33e5.png)

I am trying when I run the `ip addr` command to get the IP to use in the `cockrouch start` command, I get another different IP address.

```
ubuntu@ip-172-31-31-240:~$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc fq_codel state UP group default qlen 1000
    link/ether 0a:e5:4e:10:92:43 brd ff:ff:ff:ff:ff:ff
    inet 172.31.31.240/20 metric 100 brd 172.31.31.255 scope global dynamic eth0
       valid_lft 3067sec preferred_lft 3067sec
    inet6 fe80::8e5:4eff:fe10:9243/64 scope link
       valid_lft forever preferred_lft forever
```

1. I am assuming that 172.31.31.240 is the LAN ip address, not accessible from thr web... am I right?

2. I am working out a general-purpose recipe for automating the installation of Cockrouch clusters and nodes.
I wonder what is the exact regex to grep the output of `ip addr` and the the exact IP to bind, no matther which linux distro nor which hosting provider nor the configuration of the computer where I am working in.

**Comment by leandrosardi** [comment]
# Question

Should I run this command as is? Or should I change somthing there?

```
cockroach cert create-node localhost $(hostname) --certs-dir=certs --ca-key=my-safe-directory/ca.key
```

Is `localhost` fine, or should I write the ip address I got with the `ip addr` command?

Is `$(hostname)` fine or should I replace it by something else

**Comment by leandrosardi** [comment]
Please advice on this.
This is a new server. 
I think I am doing all steps fine.

![image](https://user-images.githubusercontent.com/55877846/170111526-0347d7b7-5447-4842-802c-38019168d673.png)

**Comment by leandrosardi** [comment]
> Hi,
> 
> I would add --max-sql-memory=.25 and --cache=.25 to the command line. These should not be left undefined.
> 
> Is localhost fine, or should I write the ip address I got with the ip addr command?
> You can add as many as you need. So you can add localhost, 127.0.0.1, the IP address from the server itself, and any names you will use to contact this node.
> 
> Example from my own test cluster:
> 
> cockroach cert create-node node-1 172.28.128.21 localhost 127.0.0.1 :: --certs-dir env/certs --ca-key env/my-safe-directory/ca.key
> Is $(hostname) fine or should I replace it by something else
> Yes, this is fine. Although note that this implies running the cockroach cert create-node command on the node itself, which means copying the private ca.key file to each node, which you may wish to avoid for security reasons. It might make more sense to create node certificates on a central server then distribute them to the nodes without copying the ca.key file.
> 
> For the final question, where you're trying to initialise the cluster, it looks like you didn't include the IP address 172.31.31.240 in the node certificate when you created it. You need to add it just like the example above.
> 
> Kind Regards,
> Shaun

**Comment by leandrosardi** [comment]
I finally have my first node published on the web.

I connected it and I ran a query using the internet IP of the server.
It worked slow, but I am sure it is because the latency of the Internet.

1. I am assuming that webservers must be in the same LAN than nodes. Right?
 
2. If I have nodes on different datacenters, i have to host webservers on each region too. Right?

3. How is exactly the interaction between the webserver and the database?

**Comment by leandrosardi** [comment]
I need a function to generate GUIDs

When I run this sentence:

```sql
-- create extension to handle guids
CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
```

I am getting this error:

```
ERROR:  unimplemented: extension "uuid-ossp" is not yet supported
HINT:  You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/54516/v21.2
SQL state: 0A000
```

How can I generate GUIDs on CRDB?

**Comment by leandrosardi** [comment]
## Question

1. When I connect the database using the Ruby `pg` gem, I specify the Internet IP of the node.

```ruby
require 'pg'
require 'sequel'

s = "postgres://blackstack:bsws2022@54.144.101.4:26257/defaultdb"
DB = Sequel.connect(s)
puts DB["SELECT 'Hello CockRouchDB!' AS message"].first[:message]
# ==> Hello CockRouchDB!
```

1. Is this a good practice?

2. Is this the unique way?

**Comment by leandrosardi** [comment]
# Question

I want to run a command that runs a SQL sentence and return to bash immediatelly. 

I need this because I am writing a Ruby script for automatig the deployment of nodes,

How to do that?

**Comment by leandrosardi** [comment]
Closing this issue now.

**Comment by leandrosardi** [comment]
# Problem

One more problem:

After running this script:

```bash
sudo -S su root -c '
cd ~;
sudo mv /usr/local/bin/cockroach /usr/local/bin/cockroach.back.202206011416050300;
sudo mv cockroach-v21.2.10.linux-amd64 cockroach-v21.2.10.linux-amd64.back.202206011416050300;
curl https://binaries.cockroachdb.com/cockroach-v21.2.10.linux-amd64.tgz | tar -xz && sudo cp -i cockroach-v21.2.10.linux-amd64/cockroach /usr/local/bin/;
sudo mv /usr/local/lib/cockroach /usr/local/lib/cockroach.back.202206011416050300;
mkdir /usr/local/lib/cockroach;
yes | sudo cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos.so /usr/local/lib/cockroach/;
yes | sudo cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos_c.so /usr/local/lib/cockroach/;
sudo mv certs certs.back.202206011416050300;
sudo mv my-safe-directory my-safe-directory.back.202206011416050300;
sudo mkdir certs;
sudo mkdir my-safe-directory;
sudo cockroach cert create-ca --certs-dir=certs --ca-key=my-safe-directory/ca.key;
sudo cockroach cert create-node localhost 172.31.21.135 $(hostname) --certs-dir certs --ca-key my-safe-directory/ca.key;
sudo cockroach cert create-client root --certs-dir=certs --ca-key=my-safe-directory/ca.key;
sudo cockroach start --certs-dir=certs --store=node0011 --listen-addr=[172.31.21.135:26257](http://172.31.21.135:26257/) --http-addr=[172.31.21.135:8080](http://172.31.21.135:8080/) --join=[172.31.21.135:26257](http://172.31.21.135:26257/) --background --max-sql-memory=.25 --cache=.25;
sudo cockroach init --host=[172.31.21.135:26257](http://172.31.21.135:26257/) --certs-dir=certs;
        '

```

I get this output that is good:

```bash
mv: cannot stat '/usr/local/bin/cockroach': No such file or directory
mv: cannot stat 'cockroach-v21.2.10.linux-amd64': No such file or directory
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 78.9M  100 78.9M    0     0  28.4M      0  0:00:02  0:00:02 --:--:-- 28.4M
mv: cannot stat '/usr/local/lib/cockroach': No such file or directory
mv: cannot stat 'certs': No such file or directory
mv: cannot stat 'my-safe-directory': No such file or directory
*
* INFO: initial startup completed.
* Node will now attempt to join a running cluster, or wait for `cockroach init`.
* Client connections will be accepted after this completes successfully.
* Check the log file(s) for progress.
*
Cluster successfully initialized

```

But when running this other script,

```bash
sudo -S su root -c 'cockroach sql --host 172.31.21.135 --certs-dir certs -e "CREATE USER blackstack WITH PASSWORD 'bsws2022';"'

```

I am getting this error message:

```bash
ERROR: x509: certificate signed by unknown authority
Failed running "sql"

```

Please advice.

**Comment by leandrosardi** [comment]
## Problem

I am writing a script to install my environment automatically.

Such a script should be capabe of re-processing.
That means: if I run the script a second time, it should finsih with no issues.

If you check the screenshot, you can see I am getting an error message when I run the command `cockroach start` at a second time, and I think the connection to the database stops working.

Please advice.

![1](https://user-images.githubusercontent.com/55877846/171467920-679395e6-8e12-4acc-bec8-988bd376a570.png)

**Comment by leandrosardi** [comment]
## Question

I created a cluster on AWS with 1 single node (node0012) by running this script below.

```
sudo su

cd ~;
sudo mv /usr/local/bin/cockroach /usr/local/bin/cockroach.back.202206011506470300;
sudo mv cockroach-v21.2.10.linux-amd64 cockroach-v21.2.10.linux-amd64.back.202206011506470300;
curl https://binaries.cockroachdb.com/cockroach-v21.2.10.linux-amd64.tgz | tar -xz && sudo cp -i cockroach-v21.2.10.linux-amd64/cockroach /usr/local/bin/;
sudo mv /usr/local/lib/cockroach /usr/local/lib/cockroach.back.202206011506470300;
mkdir /usr/local/lib/cockroach;
yes | sudo cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos.so /usr/local/lib/cockroach/;
yes | sudo cp -i cockroach-v21.2.10.linux-amd64/lib/libgeos_c.so /usr/local/lib/cockroach/;
sudo mv certs certs.back.202206011506470300;
sudo mv my-safe-directory my-safe-directory.back.202206011506470300;
sudo mkdir certs;
sudo mkdir my-safe-directory;
sudo cockroach cert create-ca --certs-dir=certs --ca-key=my-safe-directory/ca.key;
sudo cockroach cert create-node localhost 172.31.92.188 $(hostname) --certs-dir certs --ca-key my-safe-directory/ca.key;
sudo cockroach cert create-client root --certs-dir=certs --ca-key=my-safe-directory/ca.key;
sudo cockroach start --certs-dir=certs --store=node0012 --listen-addr=[172.31.92.188:26257](http://172.31.92.188:26257/) --http-addr=[172.31.92.188:8080](http://172.31.92.188:8080/) --join=[172.31.92.188:26257](http://172.31.92.188:26257/) --background --max-sql-memory=.25 --cache=.25;
sudo cockroach init --host=[172.31.92.188:26257](http://172.31.92.188:26257/) --certs-dir=certs;

sudo -S su root -c 'cockroach sql --host 172.31.92.188 --certs-dir certs -e "CREATE USER blackstack WITH PASSWORD 'bsws2022';"'

sudo -S su root -c 'cockroach sql --host 172.31.92.188 --certs-dir certs -e "CREATE DATABASE blackstack
WITH
--OWNER = blackstack
ENCODING = 'UTF8'
--LC_COLLATE = 'Spanish_Argentina.1252'
--LC_CTYPE = 'Spanish_Argentina.1252'
--TABLESPACE = pg_default
CONNECTION LIMIT = -1;"
'

sudo -S su root -c 'cockroach sql --host 172.31.92.188 --certs-dir certs -e "GRANT ALL ON DATABASE blackstack TO blackstack;"'

sudo -S su root -c 'cockroach sql --host 172.31.92.188 --certs-dir certs -e "SHOW GRANTS ON DATABASE blackstack;"'
```

Now, I created a new server on AWS/EC2 (node0014).
What script should I run in order to create a second node of the cluster above?

## Answer

Hi Leandro,

Thank you for contacting us here at Cockroach Labs! Happy to assist you with your question.

If you already have an existing cluster, you can add a node by using the cockroach start command and using the --join flag to add the node to the same addresses you used when starting the cluster.

An example of this can be found in our documentation here: [cockroach start - Add a node to a cluster](https://www.cockroachlabs.com/docs/stable/cockroach-start.html#add-a-node-to-a-cluster):

cockroach start --certs-dir=certs --advertise-addr=<node4 address> --join=<node1 address>,<node2 address>,<node3 address> --cache=.25 --max-sql-memory=.25
If you're looking to initialize a cluster with several nodes, you can do that as well (see: [Start a multi-node cluster](https://www.cockroachlabs.com/docs/stable/cockroach-start.html#start-a-multi-node-cluster)). Notice how in the example documentation, we are utilizing three cockroach start commands, each with their own IP address in the --advertise-addr flag, and joining with the other nodes using the --join flag.

Removing nodes is a bit of a longer process, as it will require both Draining and Decommissioning the node. While I would advise reading the entire page, you can utilize the information noted in the [Prepare for a graceful shutdown](https://www.cockroachlabs.com/docs/v22.1/node-shutdown.html?#prepare-for-graceful-shutdown) and [Perform node shutdown](https://www.cockroachlabs.com/docs/v22.1/node-shutdown.html?#prepare-for-graceful-shutdown) sections of the documentation to remove your desired node.

Hopefully this information helps! Certainly feel free to reach back out if you have any additional questions.

Sincerest Regards,

Nigel J. Navarro | Technical Support Engineer | Cockroach Labs

**Comment by leandrosardi** [comment]
##Problem

I have a cluster with just 1 node.

After I rebooted the server where such a node is running, I cannot longer connect the database (as you can see on screenshots 1.png and 1b.png).

Instead, I have to access the server and run again the commands `cockrach start` and `cockreach init` (as you can see in 2.png).

But after I started again the node, I noticed that all the schema was lost !!!! 

So, I had to run the installation sql scripts again (as you can see in 3.png).

Please advice!!

![1](https://user-images.githubusercontent.com/55877846/171629675-a3fda9e2-da0e-4540-bb70-ab9b5715cb2a.png)
![1b](https://user-images.githubusercontent.com/55877846/171629682-1941d7e9-4da4-4002-8298-1fdd2079fd52.png)
![2](https://user-images.githubusercontent.com/55877846/171629685-792b9719-a7e9-46bc-9bbc-0030ca530aca.png)
![3](https://user-images.githubusercontent.com/55877846/171629688-5568be25-8b90-4c36-b0e3-0a3b0f0098e3.png)

**Comment by leandrosardi** [comment]
## Question

If I run the script for installing and run CRDB for a second time, the script fails, and the CRDB stops working also.

## Answer

Thank you for the detailed explanation you have provided thus far. The process documented in our site was created as a guidance to deploy a cluster. It was not intended to be copied and pasted into a file and run it as script. There are certain steps someone has to take in order to re-install CockroachDB again on the same node depending on what you are trying to accomplish. For example, if you are trying to upgrade an existing node, the steps would be different than if you were to install a new node fresh. If you want to be able to run the script again and not install CockroachDB if it is already installed, the logic has to be written by you within the script itself. CockroachDB doesn't do this for you automatically, so you have to write a script that will check.
The best method to do what you are trying to accomplish is to use a tool like Ansible, which does exactly what you are trying to accomplish, however, if you are not familiar with it, you will have to write a script that does this instead.
In order to understand this, you must first understand what happens when you install CockroachDB. When you install CRDB, it creates a data directory and when the service starts, it listens on its configured ip address and binds to a couple of ports. Once all nodes are created, you then must connect to one of the nodes and initialize the cluster. Initializing the cluster can be done only once, so if you re-run your script again and you don't have this logic built in, the script will of course fail, since the cluster was already initialized.
Another reason your script will fail if you run it again on a node that has already been started is the fact the data directory already exists, the CockroachDB process is already running and the ports are already bound (you can't bind to the same port multiple times), so your script will also have to take this into account and you need to write this logic into your script (if you want to re-install, your script has to do certain things and you need to tell it what to do, if you want to not re-install, the script needs to know what to do). To summarize, if you want to have a script that's idempotent, you have to write it that way.
First you need to figure out what you want the script to do. Do you want it to install CockroachDB whenever it runs? Or do you want it to not do that? If you want the script to not re-install CRDB if it is already installed, then you need to write in the script to check if the process is running, if it is, then exit the script.
If you want the script to re-install CRDB, then this gets more involved, as you will have to tell the script to kill the CRDB process and remove the data directory before proceeding. You also need to tell the script to not initialize the cluster again, unless you are re-installing it in all nodes, so you also have to build this logic in the script. You also have to tell the script to check if the cluster has already been initialized, as this will fail, so this check needs to be written in the script.

With all that being said, here is the answer to your latest question and attempt in this ticket:

I am getting this error message:

ERROR: x509: certificate signed by unknown authority
Failed running "sql"

This is because your script failed when it tried to create the certs and move the binaries in place. You have to write failsafe checks in the script after each of the steps to find out where it is failing. The output from your script states the following, which is actually not good, the script failed to do all steps below, but since you did not build a failsafe in your script, it went past these errors and tried installing CRDB again:

mv: cannot stat '/usr/local/bin/cockroach': No such file or directory
mv: cannot stat 'cockroach-v21.2.10.linux-amd64': No such file or directory
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 78.9M  100 78.9M    0     0  28.4M      0  0:00:02  0:00:02 --:--:-- 28.4M
mv: cannot stat '/usr/local/lib/cockroach': No such file or directory
mv: cannot stat 'certs': No such file or directory
mv: cannot stat 'my-safe-directory': No such file or directory
My recommendation would be to re-write your script with if else statements and checks after each step, as this will help you determine where the script is failing and why.

We believe there is no further action for this ticket and have set the status to solved. If you have additional questions please reply to this ticket to reopen. If you have new issues, please open a new ticket so that can be addressed accordingly.
